src/EventListener/LocaleListener.php

```diff
@@ -51,16 +51,35 @@ public function onKernelRequest(GetResponseEvent $event)
         }
 
         $request = $event->getRequest();
+        $locale  = $this->getLocale($request);
 
-        if ($request->attributes->has('_locale')) {
-            $locale = $this->formatLocaleId($request->attributes->get('_locale'));
-        } elseif (null !== ($session = $request->getSession()) && $session->has('_locale')) {
-            $locale = $session->get('_locale');
-        } else {
-            $locale = $request->getPreferredLanguage($this->availableLocales);
+        $request->attributes->set('_locale', $locale);
+
+        if ($request->hasSession()) {
+            $request->getSession()->set('_locale', $locale);
+        }
+    }
+
+    /**
+     * Returns the locale from the request, the session or the HTTP header.
+     *
+     * @param Request $request The request object
+     *
+     * @return string The locale
+     */
+    private function getLocale(Request $request)
+    {
+        $attributes = $request->attributes;
+
+        if ($attributes->has('_locale') && null !== $attributes->get('_locale')) {
```

Comment:
I think this can be reduced to

```php
if (null !== $attributes->get('_locale')) {
```

because `ParameterBag::get()` retuns `null` if the key doesn’t exist.